File: backend/rag_chatbot/ingest_content.py
import os
import glob
from pathlib import Path
from typing import Dict, List
import hashlib

from utils.text_processing import extract_text_from_mdx, chunk_text
from services.rag_service import RAGService


def get_all_mdx_files(docs_dir: str) -> List[str]:
    """
    Get all MDX files from the docs directory.

    Args:
        docs_dir: Path to the docs directory

    Returns:
        List of paths to MDX files
    """
    # First, let's check if the directory exists
    if not os.path.exists(docs_dir):
        print(f"Directory does not exist: {docs_dir}")
        return []

    # Check what's in the directory
    print(f"Contents of {docs_dir}:")
    for root, dirs, files in os.walk(docs_dir):
        level = root.replace(docs_dir, '').count(os.sep)
        indent = ' ' * 2 * level
        print(f"{indent}{os.path.basename(root)}/")
        subindent = ' ' * 2 * (level + 1)
        for file in files:
            print(f"{subindent}{file}")

    # Look for both .mdx and .md files (Docusaurus uses both)
    mdx_pattern = os.path.join(docs_dir, "**", "*.mdx")
    mdx_files = glob.glob(mdx_pattern, recursive=True)

    md_pattern = os.path.join(docs_dir, "**", "*.md")
    md_files = glob.glob(md_pattern, recursive=True)

    print(f"Found {len(mdx_files)} .mdx files and {len(md_files)} .md files")

    # Return both types
    return mdx_files + md_files


def generate_document_id(file_path: str, content: str) -> str:
    """
    Generate a unique document ID based on file path and content.

    Args:
        file_path: Path to the document
        content: Content of the document

    Returns:
        Unique document ID
    """
    # Using UUID to ensure Qdrant compatibility
    import uuid
    doc_id = str(uuid.uuid4())
    return doc_id


def process_book_content(docs_dir: str, rag_service: RAGService):
    """
    Process all book content and add it to the knowledge base.

    Args:
        docs_dir: Path to the docs directory
        rag_service: RAG service instance to add documents to
    """
    print(f"Processing book content from: {docs_dir}")

    mdx_files = get_all_mdx_files(docs_dir)
    print(f"Found {len(mdx_files)} MDX files")

    for file_path in mdx_files:
        print(f"Processing: {file_path}")

        # Read the content of the MDX file
        with open(file_path, 'r', encoding='utf-8') as f:
            mdx_content = f.read()

        # Extract plain text from MDX
        plain_text = extract_text_from_mdx(mdx_content)

        # Break down file path to get chapter/section info
        relative_path = os.path.relpath(file_path, docs_dir)
        path_parts = relative_path.split(os.sep)
        chapter = path_parts[0] if len(path_parts) > 1 else "intro"
        section = Path(file_path).stem  # filename without extension

        # Create metadata
        metadata = {
            "title": section.replace('-', ' ').title(),
            "chapter": chapter,
            "section": section,
            "source_file": relative_path,
            "processed_at": "2025-12-17T00:00:00Z"
        }

        # Chunk the text if it's too long
        text_chunks = chunk_text(plain_text, chunk_size=1000, overlap=100)

        print(f"  - Generated {len(text_chunks)} chunks")

        # Add each chunk to the knowledge base
        for i, chunk in enumerate(text_chunks):
            # Create a unique ID for each chunk
            chunk_content = f"{chunk}"
            doc_id = generate_document_id(f"{file_path}_chunk_{i}", chunk_content)

            # Add to knowledge base
            try:
                rag_service.add_document_to_knowledge_base(
                    doc_id=doc_id,
                    content=chunk,
                    metadata={**metadata, "chunk_index": i, "total_chunks": len(text_chunks)}
                )
                print(f"    - Added chunk {i+1}/{len(text_chunks)} with ID: {doc_id[:8]}...")
            except Exception as e:
                print(f"    - Error adding chunk {i+1}: {str(e)}")

    print("Finished processing all content")


if __name__ == "__main__":
    # Initialize the RAG service
    rag_service = RAGService()

    # Process the book content - use the mounted path in the container
    docs_directory = "/app/docs"  # Path where docs are mounted in the container
    process_book_content(docs_directory, rag_service)

File: rag_chatbot_nextjs/scripts/ingest-content.ts
import { CohereClient } from 'cohere-ai';
import { Pinecone } from '@pinecone-database/pinecone';
import fs from 'fs';
import path from 'path';
import matter from 'gray-matter';
import 'dotenv/config'; // Load environment variables

// Verify environment variables exist
if (!process.env.COHERE_API_KEY) {
  throw new Error('Missing required environment variable: COHERE_API_KEY');
}

if (!process.env.PINECONE_API_KEY) {
  throw new Error('Missing required environment variable: PINECONE_API_KEY');
}

if (!process.env.PINECONE_INDEX_NAME) {
  throw new Error('Missing required environment variable: PINECONE_INDEX_NAME');
}

// Initialize services
const cohere = new CohereClient({
  token: process.env.COHERE_API_KEY!,
});

const pinecone = new Pinecone({
  apiKey: process.env.PINECONE_API_KEY!,
});

const index = pinecone.Index(process.env.PINECONE_INDEX_NAME!);

// Function to extract text from MDX content
function extractTextFromMdx(content: string): string {
  // Remove JSX components and markdown syntax
  return content
    .replace(/\{[^}]*\}/g, '') // Remove JSX expressions
    .replace(/\[(.+?)\]\(.+?\)/g, '$1') // Remove markdown links
    .replace(/!\[.*?\]\(.+?\)/g, '') // Remove images
    .replace(/^#{1,6}\s*(.*?)$/gm, '$1') // Remove headers
    .replace(/\*\*?(.+?)\*\*?/g, '$1') // Remove bold/italic
    .replace(/`(.+?)`/g, '$1') // Remove inline code
    .replace(/^>\s+(.*?)$/gm, '$1') // Remove blockquotes
    .replace(/```[\s\S]*?```/g, '') // Remove code blocks
    .replace(/^---[\s\S]*?---/m, '') // Remove frontmatter
    .replace(/\s+/g, ' ') // Normalize whitespace
    .trim();
}

// Function to chunk text
function chunkText(text: string, chunkSize: number = 1000, overlap: number = 100): string[] {
  if (text.length <= chunkSize) {
    return [text];
  }

  const chunks = [];
  let start = 0;

  while (start < text.length) {
    let end = start + chunkSize;

    // Make sure we don't go beyond the text
    if (end > text.length) {
      end = text.length;
    }

    chunks.push(text.slice(start, end));
    start = end - overlap;

    // If the remaining text is less than chunkSize, take it all
    if (text.length - start < chunkSize) {
      if (start < text.length) {
        chunks.push(text.slice(start));
      }
      break;
    }
  }

  return chunks.filter(chunk => chunk.trim().length > 0);
}

// Recursive function to get all MDX files
function getAllMdxFiles(dir: string, files: string[] = []): string[] {
  const items = fs.readdirSync(dir);

  for (const item of items) {
    const fullPath = path.join(dir, item);
    const stat = fs.statSync(fullPath);

    if (stat.isDirectory()) {
      getAllMdxFiles(fullPath, files);
    } else if (item.endsWith('.mdx')) {
      files.push(fullPath);
    }
  }

  return files;
}

// Main function to process the book content
export async function processBookContent(docsDir: string): Promise<void> {
  console.log(`Processing book content from: ${docsDir}`);

  const mdxFiles = getAllMdxFiles(docsDir);
  console.log(`Found ${mdxFiles.length} MDX files`);

  for (const file of mdxFiles) {
    console.log(`Processing: ${file}`);

    // Read the content of the MDX file
    const mdxContent = fs.readFileSync(file, 'utf8');
    const parsed = matter(mdxContent);
    
    // Extract plain text from MDX
    const plainText = extractTextFromMdx(parsed.content);

    // Break down file path to get chapter/section info
    const relativePath = path.relative(docsDir, file);
    const pathParts = relativePath.split(path.sep);
    const chapter = pathParts[0] || "intro";
    const section = path.basename(file, '.mdx'); // filename without extension

    // Create metadata
    const metadata = {
      title: section.replace('-', ' ').replace(/\b\w/g, l => l.toUpperCase()),
      chapter: chapter,
      section: section,
      source_file: relativePath,
      processed_at: new Date().toISOString()
    };

    // Chunk the text if it's too long
    const textChunks = chunkText(plainText, 1000, 100);
    console.log(`  - Generated ${textChunks.length} chunks`);

    // Add each chunk to Pinecone
    for (let i = 0; i < textChunks.length; i++) {
      const chunk = textChunks[i];
      const chunkId = `${file}_chunk_${i}`;

      // Generate embeddings for the content with rate limiting
      const embedResponse = await cohere.embed({
        texts: [chunk],
        model: 'embed-english-v3.0', // Cohere's latest embedding model
        inputType: 'search_document',
      });

      // Extract the embedding vector properly
      let embedding: number[];
      const embeddings = embedResponse.embeddings;

      if (Array.isArray(embeddings) && embeddings.length > 0 && Array.isArray(embeddings[0])) {
        // If embeddings is an array of arrays, take the first one
        embedding = embeddings[0] as number[];
      } else {
        // Fallback case
        throw new Error('Invalid embedding format received from Cohere');
      }

      // Create the document ID and metadata
      const docId = `${file}_chunk_${i}`;
      const docMetadata = {
        ...metadata,
        chunk_index: i,
        total_chunks: textChunks.length,
        content: chunk, // Store the actual content
        content_preview: chunk.substring(0, 200) + '...', // Also store a preview of the content
      };

      // Upsert the document in Pinecone
      await index.upsert([
        {
          id: docId,
          values: embedding,
          metadata: docMetadata
        }
      ]);

      console.log(`    - Added chunk ${i + 1}/${textChunks.length} with ID: ${docId}`);

      // Add a delay to respect rate limits (especially important for free tier)
      // Free tier allows 40 requests per minute, so approximately 1.5 second delay between requests
      if (i < textChunks.length - 1) {
        await new Promise(resolve => setTimeout(resolve, 1500));
      }
    }
  }

  console.log('Finished processing all content');
}

// If running this script directly
if (require.main === module) {
  const docsDirectory = process.argv[2] || '../../../my-book-website/docs'; // Default path relative to this script
  
  processBookContent(docsDirectory)
    .then(() => {
      console.log('Content ingestion completed successfully!');
      process.exit(0);
    })
    .catch((error) => {
      console.error('Error during content ingestion:', error);
      process.exit(1);
    });
}